### PR TITLE
Add documentation for e2e.Suite

### PR DESCRIPTION
**Pull request by ogaca-dd**
### What does this PR do?

This PR adds the documentation for the E2E test framework.
To generate the doc, run `pkgsite` (`go get golang.org/x/pkgsite/cmd/pkgsite@latest`) and browse `http://localhost:8080/github.com/DataDog/datadog-agent/test/new-e2e@v0.0.0/utils/e2e`

The documentation of test-infra-definition especially the VM and the Agent options must be improved.

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature. 
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

File: test/new-e2e/utils/e2e/e2e.go [hunk 1]
```diff
@@ -3,8 +3,218 @@
...
+// In our example, [e2e.EC2VMStackDef] returns an instance of [*e2e.StackDefinition][[e2e.VMEnv]].
+//
+// The following default stack definitions are provided:
+//   - [e2e.EC2VMStackDef] creates an environment with a virtual machine. See [e2e.EC2VMStackDef] for more information about the supported options.
+//   - [e2e.AgentStackDef] creates an environment with an Agent installed on a virtual machine. See [e2e.AgentStackDef] for more information about the supported options.
+//
+// # Defining your stack definition
+//
+// In some special cases, you have to define a custom environment.
+// Here is an example on how to define an environment with Docker installed on a virtual machine.
```
Comment (pducolin) [inline test/new-e2e/utils/e2e/e2e.go:87]: **🥜 nitpick**
```suggestion
// Here is an example of an environment with Docker installed on a virtual machine.
```

File: test/new-e2e/utils/e2e/e2e.go [hunk 2]
```diff
@@ -3,8 +3,218 @@
...
+//	func (suite *singleEnvSuite) Test3() {
+//		// Check feature 3
+//	}
+//
+// # Having different environments
+//
+// In this scenario, the environment is different for each test (or for most of them).
+// [e2e.Suite.UpdateEnv] is used to update the environment.
+// Keep in mind that using [e2e.Suite.UpdateEnv] to update virtual machine settings can destroy
+// the current virtal machine and create a new one when updating the operating system for example.
```
Comment (pducolin) [inline test/new-e2e/utils/e2e/e2e.go:155]:
```suggestion
// the current virtual machine and create a new one when updating the operating system for example.
```

We should also mention that it is not recommended

File: test/new-e2e/utils/e2e/e2e.go [hunk 3]
```diff
@@ -55,6 +265,21 @@ type suiteConstraint[Env any] interface {
...
 
+// Run run the tests defined in e2eSuite
+//
+// t is an instance of type [*testing.T].
+//
+// e2eSuite is a pointer to a structure with a [e2e.Suite] embbeded struct.
+//
+// stackDef defines the stack definition.
+//
+// options is an optional list of options like [DevMode], [SkipDeleteOnFailure] or [WithStackName].
```
Comment (pducolin) [inline test/new-e2e/utils/e2e/e2e.go:276]: **❓ question**
Not related to this PR: can we have a type for this options and have the full list ?

**Review by pducolin — APPROVED** [review]
Great PR! Few nits and a question on proper name for stack def / environment def

File: test/new-e2e/utils/e2e/e2e.go (hunk 3 above)
Comment (ogaca-dd) [reply test/new-e2e/utils/e2e/e2e.go:276]: I don't think so.

File: test/new-e2e/utils/e2e/e2e.go (hunk 2 above)
Comment (ogaca-dd) [reply test/new-e2e/utils/e2e/e2e.go]: I don't think so as there is a valid use case.

File: test/new-e2e/utils/e2e/e2e.go [hunk 4]
```diff
@@ -3,8 +3,218 @@
...
+//		e2e.Run(t, &vmSuite{}, e2e.EC2VMStackDef())
+//	}
+//
+//	func (v *vmSuite) TestBasicVM() {
+//		v.Env().VM.Execute("ls")
+//	}
+//
+// Writing a E2E test requires several steps.
+//
+// 1. Define your own type with the embedded [e2e.Suite] struct.
```
[the remark this answers is not in the document]
Comment (ogaca-dd) [reply test/new-e2e/utils/e2e/e2e.go:32]: I have instead [suite].

File: test/new-e2e/utils/e2e/e2e.go [hunk 5]
```diff
@@ -3,8 +3,218 @@
...
+//
+//	type vmSuite struct {
+//		e2e.Suite[e2e.VMEnv]
+//	}
+//
+// [e2e.VMEnv] defines the components available in your environment. See "Using existing stack definition section" for more information.
```
[the remark this answers is not in the document]
Comment (ogaca-dd) [reply test/new-e2e/utils/e2e/e2e.go:38]: Updated to `defines the components available in your stack`.

File: test/new-e2e/utils/e2e/e2e.go [hunk 6]
```diff
@@ -184,6 +433,7 @@ func createEnv[Env any](suite *Suite[Env], stackDef *StackDefinition[Env]) (*Env
 	return env, stackOutput, err
 }
 
+// UpdateEnv update the current environment.
```
Comment (Kaderinho) [inline test/new-e2e/utils/e2e/e2e.go:436]:
```suggestion
// UpdateEnv updates the current environment.
```

File: test/new-e2e/utils/e2e/stack_definition.go [hunk 7]
```diff
@@ -23,6 +23,7 @@ func NewStackDef[Env any](envFactory func(ctx *pulumi.Context) (*Env, error), co
 	return &StackDefinition[Env]{envFactory: envFactory, configMap: configMap}
 }
 
+// EnvFactoryStackDef create a custom stack definition
```
Comment (Kaderinho) [inline test/new-e2e/utils/e2e/stack_definition.go:26]:
```suggestion
// EnvFactoryStackDef creates a custom stack definition
```

**Review by Kaderinho — COMMENTED** [review]
Thanks for this documentation, really nicely done and very useful 🔥 🙏 
Have a couple of "broken" anchors. Please ignore if these are because I ran `pkgsite` locally.

Small nitpicks/questions:

- StackDefinition (`github.com/DataDog/datadog-agent/test/new-e2e@v0.0.0/utils/e2e`) fields are not shown because they are not exported. Imho, if technically possible, it might bring value to show them in the doc
- Same for Installer (`test-infra-definitions@main/components/datadog/agent#Installer`)
- I might have missed that, but is there any documentation on how to execute the same tests with different OS or Agent version. If I correctly understood `UpdateEnv`, it should be technically possible but what would be the most efficient way to do it ? Follow the setup describes in the `Having few environments` section ? Is there something like [GH actions matrix](https://docs.github.com/en/actions/using-jobs/using-a-matrix-for-your-jobs#using-a-matrix-strategy) ?

File: test/new-e2e/utils/e2e/e2e.go [hunk 8]
```diff
@@ -3,8 +3,220 @@
...
+//
+// To write an E2E test:
+//
+// 1. Define your own [suite] type with the embedded [e2e.Suite] struct.
+//
+//	type vmSuite struct {
+//		e2e.Suite[e2e.VMEnv]
+//	}
+//
+// [e2e.VMEnv] defines the components available in your stack. See "Using existing stack definition section" for more information.
```
[the remark this answers is not in the document]
Comment (ogaca-dd) [reply test/new-e2e/utils/e2e/e2e.go:38]: I didn't find a way to do it unfortunately.

File: test/new-e2e/utils/e2e/e2e.go [hunk 9]
```diff
@@ -3,8 +3,220 @@
...
+//	func (v *vmSuite) TestBasicVM() {
+//		v.Env().VM.Execute("ls")
+//	}
+//
+// [suite.Env] gives access to the components in your environment.
+//
+// Depending on your stack definition, [e2e.Suite.Env] can provide the following objects:
+//   - [client.VM]: A virtual machine where you can execute commands.
+//   - [client.Agent]: A struct that provides methods to run datadog agent commands.
+//   - [client.Fakeintake]: A struct that provides methods to run queries to a fake instance of Datadog intake.
```
[the remark this answers is not in the document]
Comment (ogaca-dd) [reply test/new-e2e/utils/e2e/e2e.go:63]: It was introduced by https://github.com/DataDog/datadog-agent/pull/17301#discussion_r1205273963. The type `Fakeintake` is defined at https://github.com/DataDog/datadog-agent/pull/17240/files#diff-7110dfb0c1c61f255eac565d2c84ecd1bb9b5766d96be78da0457a9a2ffbfc5a which is not yet merged. I think we can keep the link as it should work when https://github.com/DataDog/datadog-agent/pull/17240 will be merged

File: test/new-e2e/utils/e2e/e2e.go [hunk 10]
```diff
@@ -3,8 +3,220 @@
...
+//		}, nil
+//	}
+//
+//	func (docker *dockerSuite) TestDocker() {
+//		docker.Env().VM.Execute("docker container ls")
+//	}
+//
+// [e2e.EnvFactoryStackDef] is used to define a custom environment.
+// Here is a non exhaustive list of components that can be used to create a custom environment:
+//   - [EC2 VM]: Provide methods to create a virtual machine on EC2.
```
[the remark this answers is not in the document]
Comment (ogaca-dd) [reply test/new-e2e/utils/e2e/e2e.go:121]: It works now as https://github.com/DataDog/test-infra-definitions/pull/120 was merged.

File: test/new-e2e/utils/e2e/e2e.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (ogaca-dd) [reply test/new-e2e/utils/e2e/e2e.go]: Text updated: `section` is outside quotes.

File: test/new-e2e/utils/e2e/e2e.go [hunk 11]
```diff
@@ -3,8 +3,220 @@
...
+//
+//	func TestVMSuite(t *testing.T) {
+//		e2e.Run(t, &vmSuite{}, e2e.EC2VMStackDef())
+//	}
+//
+// The first argument of [e2e.Run] is an instance of type [*testing.T].
+//
+// The second argument is a pointer to an empty instance of the previous defined structure (&vmSuite{} in our example)
+//
+// The third parameter defines the environment. See "Using existing stack definition section" for more information about a environment definition.
```
[the remark this answers is not in the document]
Comment (ogaca-dd) [reply test/new-e2e/utils/e2e/e2e.go]: Text updated: `section` is outside quotes

File: test/new-e2e/utils/e2e/e2e.go [hunk 12]
```diff
@@ -3,8 +3,220 @@
...
+//
+// 3. Write a test function
+//
+//	func (v *vmSuite) TestBasicVM() {
+//		v.Env().VM.Execute("ls")
+//	}
+//
+// [suite.Env] gives access to the components in your environment.
```
[the remark this answers is not in the document]
Comment (ogaca-dd) [reply test/new-e2e/utils/e2e/e2e.go]: Link fixed.

File: test/new-e2e/utils/e2e/e2e.go [hunk 13]
```diff
@@ -101,21 +332,33 @@ func SkipDeleteOnFailure[Env any]() func(*Suite[Env]) {
 
 // Env returns the current environment.
 // In order to improve the efficiency, this function behaves as follow:
-//   - It creates the default environment if no environment exists. It happens only during the first call of the test suite.
-//   - It restores the default environment if UpdateEnv was not already called during this test.
-//     This avoid having to restore the default environment for each test even if UpdateEnv immedialy
-//     overrides this environment.
+//   - It creates the default environment if no environment exists.
+//   - It restores the default environment if [suite.UpdateEnv] was not already called during this test.
```
[the remark this answers is not in the document]
Comment (ogaca-dd) [reply test/new-e2e/utils/e2e/e2e.go]: Link fixed.

**Review by Kaderinho — APPROVED** [review]

File: test/new-e2e/utils/e2e/e2e.go [hunk 14]
```diff
@@ -3,8 +3,220 @@
...
+//		docker.Env().VM.Execute("docker container ls")
+//	}
+//
+// [e2e.EnvFactoryStackDef] is used to define a custom environment.
+// Here is a non exhaustive list of components that can be used to create a custom environment:
+//   - [EC2 VM]: Provide methods to create a virtual machine on EC2.
+//   - [Agent]: Provide methods to install the Agent on a virtual machine
+//   - [File Manager]: Provide methods to manipulate files and folders
+//
+// # Organizing your tests
```
Comment (vboulineau) [inline test/new-e2e/utils/e2e/e2e.go:125]: I'd add a **NOTE** (in bold) to emphasize that `TestXX` on `Suite` are **not** executed in top->bottom order (unlike normal go tests) and that test ordering is random and not guaranteed.

**Review by vboulineau — APPROVED** [review]

File: test/new-e2e/utils/e2e/e2e.go (hunk 3 above)
Comment (ogaca-dd) [reply test/new-e2e/utils/e2e/e2e.go:281]: As there are functions I don't think so.

**Pull request merged**